docs/proposals/volumes.md

```diff
@@ -0,0 +1,431 @@
+volume could affect changes in the state of arbitrary paths within the host's filesystem.  This
+would be a severe security risk, so we will consider hostPath a corner case that the kubelet should
+never perform ownership management for.
+
+#### Shared storage: network
+
+Ownership management of shared storage is a complex topic.  Ownership for existing shared storage
+will be managed externally from Kubernetes.  For this case, our API should make it simple to express
+whether a particular volume should have these concerns managed by Kubernetes.
+
+We will not attempt to address the actual ownership and permissions concerns of new shared storage
+in this proposal.
+
+#### API requirements
+
+From the above, we know that ownership management must be applied:
+
+1.  To some volume types always
+2.  To some volume types never
+3.  To some volume types *sometimes*
+
+For volume types where the need for ownership/label management is not at the discretion of the user,
+there should be no need (or way) to specify whether ownership management is required in the API.
+Conversely there must be a way to specify this for volume types where it depends on the user's
+needs.
+
+### Ownership management
+
+Using the approach of a pod-level supplemental group to own volumes solves the problem in any of the
+cases of UID/GID combinations within a pod. Since this is the simplest approach that handles all
+use-cases, our solution will be made in terms of it.
+
+Eventually, Kubernetes should allocate a unique group for each pod so that a pod's volumes are
+usable by that pod's containers, but not by containers of another pod.  The supplemental group used
+to share volumes must be unique in a multitenant cluster.  If uniqueness is enforced at the host
+level, pods from one host may be able to use shared filesystems meant for pods on another host.
+
+An admission controller could handle allocating groups for each pod and setting the group in the
+pod's security context.
+
+#### When to manage ownership
+
+Whether Kubernetes should manage the ownership of a volume for a volume depends upon
+both the volume type and the cluster operator's policy.  For example:
+
+1.  For shared storage, ownership of volumes externally to the cluster
+2.  It is not possible to securely `chown` or `chmod` paths within some shared storage
+
+The system should be designed so that ownership management is only applied to the appropriate
+volumes.
+
+### Setting ownership and permissions on volumes
+
+For volumes that are created by the system, it is sufficient to perform `chown` and `chmod`
+operations on the host's filesystem.  For unshared storage based on remote block devices, `chown`
+and `chmod` applied locally will also work correctly.
+
+## Proposed design:
+
+Our design should minimize code for handling ownership required in the Kubelet and volume plugins.
+
+### API changes
+
+We should not interfere with images that need to run as a particular UID or primary GID.  A pod
```

Comment:
I think you need special consideration in this proposal for the root supplemental group, which people depend on today if the uid is arbitrary (by discovery, they know that root gid works for volume permissions).  Adding an explicit supplemental groups that *replaces* GID 0 breaks those, which means users can't predict what GIDs will work in the container.  We need to make a statement that either all containers can have GID 0 as a supplemental group, or that admins can disable it, but we need to clarify.